pkg/controller/database/cosmosdb/managed.go

```diff
@@ -0,0 +1,164 @@
+	name := managed.ControllerName(v1beta1.CosmosDBAccountGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.CosmosDBAccount{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.CosmosDBAccountGroupVersionKind),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{kube: mgr.GetClient(), newClientFn: cosmosdb.NewDatabaseAccountClient}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	kube        client.Client
+	newClientFn func(creds []byte) (cosmosdb.AccountClient, error)
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return nil, errors.New(errNotNoSQLAccount)
+	}
+
+	p := &v1alpha3.Provider{}
+	n := meta.NamespacedNameOf(r.Spec.ProviderReference)
+	if err := c.kube.Get(ctx, n, p); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider %s", n)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretNil)
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider secret %s", n)
+	}
+
+	client, err := c.newClientFn(s.Data[p.Spec.CredentialsSecretRef.Key])
+	return &external{client: client}, errors.Wrap(err, "cannot create new Azure Database Account client")
+}
+
+// external is a createsyncdeleter using the Azure API.
+type external struct {
+	client cosmosdb.AccountClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotNoSQLAccount)
+	}
+
+	res, err := e.client.CheckNameExists(ctx, meta.GetExternalName(r))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetNoSQLAccount)
+	}
+
+	if res.Response.StatusCode == http.StatusNotFound {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	account, err := e.client.Get(ctx, r.Spec.ForProvider.ResourceGroupName, meta.GetExternalName(r))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetNoSQLAccount)
+	}
+	cosmosdb.UpdateCosmosDBAccountObservation(&r.Status, account)
+
+	switch r.Status.AtProvider.State {
+	case "Succeeded":
+		r.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(r)
+	default:
+		r.SetConditions(runtimev1alpha1.Unavailable())
+	}
+
+	resourceUpToDate := cosmosdb.CheckEqualDatabaseProperties(r.Spec.ForProvider.Properties, r.Status.AtProvider.Properties)
```

Comment:
It's unusual for us to maintain an identical set of fields under `status.atProvider` to fields that are configurable in `spec.forProvider`. Typically if a field can be configured by a user, it lives only under `spec`.

In this case most Kubernetes resources would treat any field with a value specified under `spec.forProvider.properties` as authoritative and "late initialize" any fields the CR author did not specify by automatically setting any omitted `spec.forProvider.properties` to the default values set by Azure (if any). There's a little more detail about our position on this in our [API design one pager](https://github.com/crossplane/crossplane/blob/3c9e53/design/one-pager-managed-resource-api-design.md#high-fidelity).

(CC @muvaf, who I think raised an issue tracking the fact that this means we risk being out of sync with the cloud provider API without letting the CR author know that.)

Reply:
Note - I'd be happy to merge this at `v1alpha1` with this pattern in place and figure out a different approach before we go to `v1beta1`.

Reply:
Will take a look at that.

Reply:
I removed the structure under Status. It was not needed. It just has the State property now.